From 06b733b4fd0e1309b11cb1446e2240c6fcadd159 Mon Sep 17 00:00:00 2001
From: nchernikov <nchernikov@maxlinear.com>
Date: Mon, 23 Jun 2025 11:54:24 +0300
Subject: [PATCH] WLANRTSYS-90131 Update RNR IEs when 6G is reconfigured with
 SIGHUP, SET or RECONF

Issue:
RNR IEs and particularly short SSID is not updated when SSID is changed in any 6G VAP

Fix:
Update RNR IEs for all the configuration flows

Fix up:
WLANRTSYS-57484: Reconf: Add ubus and reload patches
WLANRTSYS-55538: Reconf implementation

Signed-off-by: nchernikov <nchernikov@maxlinear.com>
---
 hostapd/ctrl_iface.c | 10 ++++++++++
 src/ap/hostapd.c     | 34 ++++++++++++++++++++++++++++++++++
 2 files changed, 44 insertions(+)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 68033aacc..35d9d6f1c 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -1398,6 +1398,16 @@ static int hostapd_ctrl_iface_set(struct hostapd_data *hapd, char *cmd)
 			hostapd_disassoc_accept_mac(hapd);
 		} else if (os_strcasecmp(cmd, "ssid") == 0) {
 			hostapd_neighbor_sync_own_report(hapd);
+#ifdef CONFIG_VENDOR_MXL
+			struct hostapd_iface *iface = hapd->iface;
+			if (is_6ghz_op_class(iface->conf->op_class)) {
+				if (mxl_update_rnr_check(iface)) {
+					wpa_printf(MSG_ERROR,
+						   "Failed to update RNR on ssid set");
+					return -1;
+				}
+			}
+#endif /* CONFIG_VENDOR_MXL */
 		} else if (os_strncmp(cmd, "wme_ac_", 7) == 0 ||
 			   os_strncmp(cmd, "wmm_ac_", 7) == 0
 #ifdef CONFIG_EPCS_MXL
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 4f7bae661..8de331355 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -340,6 +340,15 @@ int hostapd_reload_config(struct hostapd_iface *iface)
 		if (res < 0)
 			wpa_printf(MSG_ERROR,
 				   "Failed to enable interface on config reload");
+#ifdef CONFIG_VENDOR_MXL
+		if (is_6ghz_op_class(iface->conf->op_class)) {
+			if (mxl_update_rnr_check(iface)) {
+				wpa_printf(MSG_ERROR,
+					   "Failed to update RNR after force full disable+enable sequence");
+				return -1;
+			}
+		}
+#endif /* CONFIG_VENDOR_MXL */
 		return res;
 	}
 
@@ -406,6 +415,15 @@ int hostapd_reload_config(struct hostapd_iface *iface)
 	iface->conf = newconf;
 	hostapd_config_free(oldconf);
 
+#ifdef CONFIG_VENDOR_MXL
+		if (is_6ghz_op_class(iface->conf->op_class)) {
+			if (mxl_update_rnr_check(iface)) {
+				wpa_printf(MSG_ERROR,
+					   "Failed to update RNR after config reload");
+				return -1;
+			}
+		}
+#endif /* CONFIG_VENDOR_MXL */
 
 	return 0;
 }
@@ -470,6 +488,14 @@ int mxl_hostapd_reload_config(struct hostapd_iface *iface)
 	res = hostapd_restart_all_changed_bss(iface, newconf);
 	hostapd_config_free(newconf);
 
+	if (is_6ghz_op_class(iface->conf->op_class)) {
+		if (mxl_update_rnr_check(iface)) {
+			wpa_printf(MSG_ERROR,
+				   "Failed to update RNR after reconf reload config");
+			return -1;
+		}
+	}
+
 	return res;
 }
 
@@ -5092,6 +5118,14 @@ int hostapd_reconf_iface(struct hostapd_iface *hapd_iface, int changed_idx)
 
 	hostapd_config_free(new_conf);
 
+	if (is_6ghz_op_class(hapd_iface->conf->op_class)) {
+		if (mxl_update_rnr_check(hapd_iface)) {
+			wpa_printf(MSG_ERROR,
+				   "Failed to update RNR after iface reconf");
+			return -1;
+		}
+	}
+
 	return 0;
 }
 
-- 
2.43.0

